### PR TITLE
Add panic handling when attaching the bpf program to do not termine kepler

### PR DESCRIPTION
**Pull request by marceloamaral**
#### Why this PR is needed?
Kepler should work in the case that it could not attach a bpf program. We have the logic for this.
However, there are sometimes that it causes a panic and then terminates the program.

The PR #275 having this problem where the VM does not have the `Module kheaders`, details are [here](https://github.com/sustainable-computing-io/kepler/actions/runs/3201921603/jobs/5230372207#step:8:13).

This PR fix the issue #282

#### What this PR does?
Golang [recover](https://gobyexample.com/recover) function make it possible to recover from a panic.
Therefore, we use this when attaching the bpf program and instead of terminate the program we just log the error.

Signed-off-by: Marcelo Amaral <marcelo.amaral1@ibm.com>

### CODE REVIEW

**Comment by SamYuan1990** [comment]
I suggest add parts below in CI to check pod status, and verify this `recover` works. (logically it should, but just in case.)
```      - name: get pod logs
        run: |
          sleep 60
          kubectl logs $(kubectl -n kepler get pods -oname) -n kepler
          kubectl get all -n kepler
```

**Comment by marceloamaral** [comment]
@rootfs please, do not merge this PR
I added this logic to test if the kepler is still alive but it has a sleep of 60s, which will make the test very slow

After verifying this, I will remove this logic

**Comment by SamYuan1990** [comment]
@marceloamaral , the pod https://github.com/sustainable-computing-io/kepler/actions/runs/3202720302/jobs/5231993693#step:7:28 is CrashLoopBackOff.

**Comment by SamYuan1990** [comment]
Attach function need a recover too?
```
goroutine 1 [running]:
github.com/iovisor/gobpf/bcc.(*Module).TableId.func2(0x2705d70?, 0x9?)
	/opt/app-root/src/github.com/sustainable-computing-io/kepler/vendor/github.com/iovisor/gobpf/bcc/module.go:467 +0x19
github.com/iovisor/gobpf/bcc.(*Module).TableId(0x15979c0?, {0xd35109?, 0x0?})
	/opt/app-root/src/github.com/sustainable-computing-io/kepler/vendor/github.com/iovisor/gobpf/bcc/module.go:467 +0x6a
github.com/sustainable-computing-io/kepler/pkg/attacher.AttachBPFAssets()
	/opt/app-root/src/github.com/sustainable-computing-io/kepler/pkg/attacher/bcc_attacher.go:128 +0x4e5
github.com/sustainable-computing-io/kepler/pkg/collector.(*Collector).Attach(0xc00040e2b0)
	/opt/app-root/src/github.com/sustainable-computing-io/kepler/pkg/collector/collector.go:104 +0x33
main.main()
	/opt/app-root/src/github.com/sustainable-computing-io/kepler/cmd/exporter.go:80 +0x40f
```

**Comment by SamYuan1990** [comment]
@marceloamaral , I noticed one thing. as https://github.com/sustainable-computing-io/kepler/actions/runs/3204057772/jobs/5234890386#step:7:16
please let me know as we installed prometheus operator, will prometheus auto invoke some url for kepler pod?

as I am worried, just catch `Attach()` function will not fix all possible issues.

**Comment by marceloamaral** [comment]
@SamYuan1990 Kepler is being successfully deployed now

The point with `recover()` is that we need to have the err variable defined in the function output, otherwise it will "loose" the content.

Anyway, I fixed that and it is working now! Please have a look, if you agree, I will remove the CI code.

**Comment by marceloamaral** [comment]
> will prometheus auto invoke some url for kepler pod?

Yes, that is how Prometheus works. It will scrape metrics from applications.

**Review by SamYuan1990 — APPROVED** [review]
LGTM

**Comment by SamYuan1990** [comment]
> @SamYuan1990 Kepler is being successfully deployed now
> 
> The point with `recover()` is that we need to have the err variable defined in the function output, otherwise it will "loose" the content.
> 
> Anyway, I fixed that and it is working now! Please have a look, if you agree, I will remove the CI code.

I hope you can keep the CI code.

File: cmd/exporter.go
Comment (SamYuan1990) [inline cmd/exporter.go:82]: @rootfs , but we may need to think twice before merge. 
for this pr, we can accept with info level for now, if bcc libs not found on host instance. for long term considering, 
Q: if host instance/vm without bcc then what should we do?
Q: should we provide any by default approach to ensure restful api works?

Comment (marceloamaral) [reply cmd/exporter.go:82]: The api will work if bpf program is not running. Which means that the code will execute without performance counters.

**Comment by SamYuan1990** [comment]
I will re-check this pr at about 24 hours ago?
if this pr removed [WIP] without merge, then I will merge it. cc: @rootfs

**Comment by marceloamaral** [comment]
@SamYuan1990 ok, let's keep the CI code.
In the case the test start to be slow, we should verify it again.

File: pkg/attacher/bcc_attacher.go
Comment (rootfs) [inline pkg/attacher/bcc_attacher.go:67]: there is no more panic or fatalf, recover() won't triggered

File: cmd/exporter.go
Comment (rootfs) [reply cmd/exporter.go:82]: the collector loop will not be able to read anything if eBPF is not loaded. That'll bring more to debug. I suggest:
- add build flags for panic()/fatalf and only build with them for CI test
- investigate non-bcc alternatives in a separate stream if necessary

Comment (SamYuan1990) [reply cmd/exporter.go:82]: > the collector loop will not be able to read anything if eBPF is not loaded. That'll bring more to debug. I suggest:
> 
> * add build flags for panic()/fatalf and only build with `Infof` for CI test
> * investigate non-bcc alternatives in a separate stream if necessary

In my point of view, I suppose this PR is just a bug fix, and we can do further thing through #282 or a new ticket.
I am not sure if #282 , related with #192  and blocks #162 as the pod in CI will not in running status before this fix.

And once again, I hope we can make kepler running in MVC pattern for restapi.
which 
- return number of 0/-1/-NaN instead of a panic as restful response.
- use adopter/strategy pattern adopt with different eBPF libraries and CPU architectures.

or we have an agreement that if kepler running on vm/instance which fails with eBPF loading there should be a panic?
if so, which means for CI we may have to install bcc at ubuntu based github action agent.

File: pkg/attacher/bcc_attacher.go
Comment (SamYuan1990) [reply pkg/attacher/bcc_attacher.go:67]: well, the 1st panic real happens at loadModule function, and the `m := bpf.NewModule(string(objProg), options)` real good as it init variable of `m`. As a hot fix, I do agree to make it as it is.
or maybe we'd better give some comments in line around or put init variable of m in creation/init function when controller been created.

Comment (marceloamaral) [reply pkg/attacher/bcc_attacher.go:67]: @rootfs Sorry I did not understand your comment...
We do have a panic happening in the case there is any problem with bpf attach
But Kepler should not stop to run if we don't have bpf

**Comment by SamYuan1990** [comment]
@rootfs are we good enough to merge this PR?

**Pull request merged**
